components/char_collection/src/char_collection.rs

```diff
@@ -0,0 +1,653 @@
+// Copyright 2019 The Fuchsia Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+use {
+    anyhow::{format_err, Error},
+    std::{
+        clone::Clone,
+        cmp::Ordering,
+        hash::{Hash, Hasher},
+        iter::Iterator,
+        ops::Range,
+        vec::Vec,
+    },
+    unic_char_range::{chars, CharIter, CharRange},
```

Comment:
It seems to me that it would be nicer to define `CharRange(RangeInclusive<char>)` with the appropriate methods, and conversion methods from/to the different range types. The macro would then be unnecessary. 

Rust ranges already cover most of what we want here, the only difference is that we need to enforce that `start <= end`.

Reply:
An alternate approach is to make the API directly operate on RangeInclusive, and assert for bad ranges on insertion

Reply:
If you decide to go for `CharRange(RangeInclusive<char>)` we could have `as_range` method that returns that as a range for use on all APIs that accept `impl RangeBounds`.